runtime/lua/vim/_options.lua

```diff
@@ -134,6 +134,119 @@ local function get_options_info(name)
   return info
 end
 
+--- @type table<string,function>
+vim._func_opts = {}
+
+-- Options that can be set as a function
+local func_opts = {
+  completefunc = true,
+  findfunc = true,
+  omnifunc = true,
+  operatorfunc = true,
+  quickfixtextfunc = true,
+  tagfunc = true,
+  thesaurusfunc = true,
+}
+
+-- Options that can be set as an expression
+local expr_opts = {
+  diffexpr = true,
+  foldexpr = true,
+  formatexpr = true,
+  includeexpr = true,
+  indentexpr = true,
+  modelineexpr = true,
+  patchexpr = true,
+}
+
+-- Options that can be set as an expression with a prefix of '%!'
+local pct_bang_expr_opts = {
+  statuscolumn = true,
+  statusline = true,
+  tabline = true,
+  winbar = true,
+}
+
+--- @param v integer?
+--- @param f fun():integer
+--- @return integer
+local function resolve_winbuf(v, f)
+  return assert((not v or v == 0) and f() or v)
+end
+
+--- @type table<string,boolean>
+local all_func_opts = vim.tbl_extend('error', func_opts, expr_opts, pct_bang_expr_opts)
+
+--- If the option name is a function option convert it to a string
+--- format and store the function value.
+--- @param name string
+--- @param value any
+--- @param info vim.api.keyset.get_option_info
+--- @param opts vim.api.keyset.option
+local function apply_func_opt(name, value, info, opts)
+  local t = nil
+  local t_str --- @type string?
+
+  -- Find a table to store/clear the function
+  if info.scope == 'global' or info.global_local and opts.scope == 'global' then
```

Comment:
@famiu instead of cleaning up functions using an autocmd, I decided to use `vim.b/w` to hold the references.

Does this monstrosity of code make sense to you?

Reply:
> @famiu instead of cleaning up functions using an autocmd, I decided to use `vim.b/w` to hold the references.
> 
> Does this monstrosity of code make sense to you?

Somehow GitHub didn't notify me about this ping.

But yeah the function makes sense to me, although it took me a little while to parse entirely. Not the worst I've seen when it comes to code related to options.